packages/core/src/sync/index.ts

```diff
@@ -566,180 +579,241 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
+              factoryLogs.map((log) => ({ log })),
+            ),
+            shouldUpdateCheckpoint: false,
+            chainId: network.chainId,
+          }),
+          args.syncStore.insertTransactions({
+            transactions: finalizedEventData.flatMap(
+              ({ transactions }) => transactions,
+            ),
             chainId: network.chainId,
+          }),
+          args.syncStore.insertTransactionReceipts({
+            transactionReceipts: finalizedEventData.flatMap(
+              ({ transactionReceipts }) => transactionReceipts,
+            ),
+            chainId: network.chainId,
+          }),
+          args.syncStore.insertCallTraces({
+            callTraces: finalizedEventData.flatMap(({ callTraces, block }) =>
+              callTraces.map((callTrace) => ({ callTrace, block })),
+            ),
+            chainId: network.chainId,
+          }),
+          ...args.sources
+            .filter(({ filter }) => filter.chainId === network.chainId)
+            .map(({ filter }) =>
+              args.syncStore.insertInterval({ filter, interval }),
+            ),
+        ]);
+
+        /**
+         * The realtime service can be killed if `endBlock` is
+         * defined has become finalized.
+         */
+        if (isSyncComplete(syncProgress)) {
+          args.common.metrics.ponder_sync_is_realtime.set(
+            { network: network.name },
+            0,
+          );
+          args.common.metrics.ponder_sync_is_complete.set(
+            { network: network.name },
+            1,
+          );
+          args.common.logger.info({
+            service: "sync",
+            msg: `Synced final end block for '${network.name}' (${hexToNumber(syncProgress.end!.number)}), killing realtime sync service`,
           });
+          // TODO(kyle) is this okay?
+          realtimeSync.kill();
+        }
+        break;
+      }
+      /**
+       * Handle a reorg with a new common ancestor block being found.
+       */
+      case "reorg": {
+        syncProgress.current = event.block;
+        const checkpoint = getOmnichainCheckpoint("current")!;
 
-          args.onRealtimeEvent({ type: "reorg", checkpoint });
+        // Update "ponder_sync_block" metric
+        args.common.metrics.ponder_sync_block.set(
+          { network: network.name },
+          hexToNumber(syncProgress.current.number),
+        );
+
+        localSyncContext.get(network)!.unfinalizedEventData =
+          unfinalizedEventData.filter(
+            (led) =>
+              hexToNumber(led.block.number) <= hexToNumber(event.block.number),
+          );
 
-          break;
+        const reorgedHashes = new Set<Hash>();
+        for (const b of event.reorgedBlocks) {
+          reorgedHashes.add(b.hash);
         }
 
-        default:
-          never(event);
+        unindexedEvents = unindexedEvents.filter(
+          (e) => reorgedHashes.has(e.block.hash) === false,
+        );
```

Comment:
I'm getting a bit confused by the difference between `unfinalizedEventData` and `unindexedEvents` here. It seems like they contain the same information, why do we need both?

Reply:
They do contain the same information, where `unindexedEvents` is derived from `unfinalizedEventData`. However, when building events, point-in-time child addresses are needed. Also, the lifetime of each of these variables is different.